### PR TITLE
[ready] Add logdet and slogdet

### PR DESCRIPTION
**Pull request by ssnl**
See commit messages for details of each commit. Notice that the last commit changes the QR based approach to LU based.

Some comments about switching from QR to LU:

1. QR (`geqrf`) is generally more stable than LU (`getrf`).
2. But `getrf` gives specific information about matrix being singular or not. Furthermore, the backward (when det > 0) uses matrix inverse (`getrf` + `getri`), so using `getrf` in forward makes sense.
3. Finally, this is also more consistent with many other libraries (including NumPy and MATLAB) that use LU factorization (`getrf`) to compute determinant. 

Therefore, in this PR, det is computed using `btrtifact` (batched `getrf`). However, the backward still uses `inverse` (`getrf` + `getri`). So future work include:
1. expose `getri` and cache `getrf` results to speed up backward.
2. implement batched `getri` to support batched det methods.

### CODE REVIEW

**Comment by vishwakftw** [comment]
@SsnL I will have to use logdet for the implementation of the KL-divergence between two multivariate normal distributions. I would like to know how fast would this be as opposed to computing the determinant and taking the log of it?

**Comment by soumith** [comment]
putting speed aside, the gradient formulas permitting this would prob be more numerically stable

**Comment by SsnL** [comment]
@zou3519 can you review this when you have time? I moved some doc changes to other PR so it's shorter now. :) Basically, maths is:

In forward: use QR factorization to get A=QR, then det(A) = det(Q) det(R). det(Q) is obtained by counting the number on nnz entries in tau (-1 if odd, and 1 if even). And det(R) is the product of diag(R).

In backward: calculate a specific variant of SVD: A = USV^T, s.t. det(UV) = 1. So det(A) = det(S) = prod(diag(S)). Then use svd_backward to backprop the determinant as if it is calculated using this method.

File: test/test_autograd.py [hunk 1]
```diff
@@ -2532,13 +2566,34 @@ class dont_convert(tuple):
...
+    # `logdet`, we also set `make_nonzero_det(matrix, sign=1)` to make the
+    # matrix have positive det.
+    ('logdet', lambda: make_nonzero_det(torch.randn(S, S), 1), NO_ARGS, '', NO_ARGS, [skipIfNoLapack]),
+    ('logdet', lambda: make_nonzero_det(random_symmetric_matrix(S), 1), NO_ARGS,
+     'symmetric', NO_ARGS, [skipIfNoLapack]),
+    ('logdet', lambda: make_nonzero_det(random_symmetric_pd_matrix(S), 1), NO_ARGS,
+     'symmetric_pd', NO_ARGS, [skipIfNoLapack]),
+    ('logdet', lambda: make_nonzero_det(random_fullrank_matrix_distinct_singular_value(S), 1, 0), NO_ARGS,
+     'distinct_singular_values', NO_ARGS, [skipIfNoLapack]),
+    ('slogdet', lambda: make_nonzero_det(torch.randn(S, S), 1), NO_ARGS, 'pos_det', NO_ARGS, [skipIfNoLapack], [1]),
```
Comment (SsnL) [inline test/test_autograd.py]: the last element of [1] means that only the 2nd output of slogdet is differentiable, and should be gradcheck'ed

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -8,50 +8,100 @@
 namespace at {
 namespace native {
 
-// For backward, we save svd.
+// This method calculates svd with an additional constraint (det(UV) = +1),
+// which will be used in det methods backward.
+//
 // http://www.ics.forth.gr/cvrl/publications/conferences/2000_eccv_SVD_jacobian.pdf
-// But instead of gesvd SVD A = U(A) Sig(A) V(A)^T, which doesn't specify signs
+// Instead of gesvd SVD A = U(A) Sig(A) V(A)^T, which doesn't specify signs
```
Comment (zou3519) [inline aten/src/ATen/native/LinearAlgebra.cpp]: What do you mean by "doesn't specify signs?"

**Comment by SsnL** [comment]
Hold on reviewing @zou3519 . I plan to move det computation from QR to LU factorization. It would be more consistent with things like numpy matlab and also won't be having problem with the inverse in backward.

**Comment by SsnL** [comment]
Some comments about switching to LU:

1. QR (`geqrf`) is generally more stable than LU (`getrf`).
2. But `getrf` gives specific information about matrix being singular or not. Furthermore, the backward (when det > 0) uses matrix inverse (`getrf` + `getri`), so using `getrf` in forward makes sense.
3. Finally, this is also more consistent with many other libraries (including NumPy and MATLAB) that use LU factorization (`getrf`) to compute determinant. 

Therefore, in this PR, det is computed using `btrtifact` (batched `getrf`). However, the backward still uses `inverse` (`getrf` + `getri`). So future work include:
1. expose `getri` and cache `getrf` results to speed up backward.
2. implement batched `getri` to support batched det methods.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (SsnL) [reply aten/src/ATen/native/LinearAlgebra.cpp]: signs of determinants of U and V. The sentence continues on the next line :). No worries. I'll ask someone with more experience to review this.

**Comment by SsnL** [comment]
@vishwakftw @samuela  do one of you want to review this PR? Thanks in advance! :)

**Review by vishwakftw — APPROVED** [review]
I think the code looks good. Thanks!!

**Comment by SsnL** [comment]
@vishwakftw @soumith Thanks!!!

**Pull request merged**